Adds backwards compatibility in `UsingPrecompile` contract.

soloseng opened the pull request:
### Description

Updated `UsingPrecompile` contract to maintain backwards compatibility after L2 migration. 

Fucntions that maintained backwards compatibility query the required information from the `EpochManager` contract instead of the precompiles.

- [ ] Remove onlyL1 in validators.sol

### Other changes

Was required to add the list of elected validators to the `Epoch` struct of the EpochManager contract to keep the historical list of elected validators at a given epoch number.

### Tested

Added unit test.

### Related issues

- Fixes https://github.com/celo-org/celo-blockchain-planning/issues/642

### Backwards compatibility

for some select functions

soloseng committed:
added `getEpochInfoOfEpoch(uint256)` to get epoch info of specific epochs

soloseng committed:
updated natspec

soloseng committed:
PR feedback

soloseng committed:
Use constant

soloseng committed:
breakup test; added more

soloseng committed:
++ registry to precompile contract

soloseng committed:
Merge branch 'soloseng/epoch-of-block' into soloseng/backward-compatibility-fix

soloseng committed:
added elected to epoch info

soloseng committed:
numberValidatorsInSet backwards compatibility

soloseng committed:
validator address & validatorSigner address from set

soloseng committed:
support `getEpochNumber()` backwards compatibility in UsingRegistry

soloseng committed:
using `getEpochByBlockNumber` instead of `getEpochByNumber`

soloseng committed:
limit span of loop only to L2 blocks

soloseng committed:
use binary search instead of blind loop

soloseng committed:
++ natspec

soloseng committed:
added getEpochNumberOfBlock() backwards compatibility support

soloseng commented:
As of 50bec5c8a89c6e7d9026818ef1d591cf0e891b53
I was able to add the required functions to preserve backwards compatibility to the UsingRegistry contract.

Testing is not fully implemented for registry functions, but the plan was to use `EpochManager_WithMock.sol` for testing needs and if possible remove `MockEpochManager.sol`

stCELO only uses `numberValidatorsInCurrentSet` ;  `validatorSignerAddressFromCurrentSet` but I also added compatibility for the functions that perform the same query using specific blocknumber incase other contracts in the CELo ecosystem use them.

To get the epoch info based on block number, I used binary search to avoid looping through all the epoch 1x1. This was required for `getEpochNumberOfBlock()`; `validatorSignerAddressFromSet()`

We could further improve the compute cost by introducing a mapping of blocksToEpoch, 
```
mapping(uint256 => uint256) internal blockToEpoch;
``` 
but that would be a more complex change and increase the gas cost of processing an epoch as we'd have to store this info during processing.



**NOTE:** Not all functions have backwards compatibility.

martinvol commented on packages/protocol/contracts/common/UsingPrecompiles.sol:
```diff
@@ -2,9 +2,11 @@ pragma solidity ^0.5.13;
 import "../../contracts-0.8/common/IsL2Check.sol";
 
-contract UsingPrecompiles is IsL2Check {
+contract UsingPrecompiles is IsL2Check, UsingRegistry {
```
UsingRegistry has a registry variable that needs to be set, if we add the dependency here then if the childs don't have UseRegistry they would break.

martinvol replied:
Updating it here would also trigger redeployments of every child

m-chrzan replied:
Can UsingRegistryV2 be used instead? That uses only `constants`. Or is the Registry address different on L2?

pahor167 replied:
Yep, unfortunately it not possible to add new inherited smart contract with storage variable to already deployed smart contract (there is very high probability that this change will break storage)

soloseng replied:
using `PrecompileOverride` contract, as discussed [here](https://github.com/celo-org/celo-blockchain-planning/issues/642#issuecomment-2397407994)

m-chrzan commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -400,6 +402,38 @@ contract EpochManager is
+   * @param _blockNumber Block number of the epoch info is retreived.
+   */
+  function getEpochNumberOfBlock(uint256 _blockNumber) external view returns (uint256) {
+    (uint256 _epochNumber, , , , , ) = _getEpochByBlockNumber(_blockNumber);
```
Is there a way to do this without calling `_getEpochByBlockNumber`? If this is expected to be used in transactions rather than only off-chain view calls (by us or any third party protocol), reading the entire `elected` array into memory will waste a good bit of gas when only the epoch number is needed.

m-chrzan commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -494,4 +555,73 @@ contract EpochManager is
+
+    while (left <= right) {
+      uint256 mid = (left + right) / 2;
+      Epoch memory _epoch = epochs[mid];
```
Would this be much more gas efficient with only grabbing a `storage` ref (or reading only `first`/`lastBlock` into memory for the binsearch)? Not sure if the compiler optimizes away reading the other fields (including the entire `elected` array) into memory, which could incur a lot of unnecessary `SLOAD`s while searching.

soloseng replied:
This struct no longer includes the election history

m-chrzan commented on packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
    * @return Address of validator at the requested index.
    */
+  function validatorAddressFromCurrentSet(uint256 index) public view onlyL2 returns (address) {
+    address[] memory _elected = getEpochManager().getElected();
```
More gas optimization: add an indexed function to EpochManager that reads only the specific index, rather than reading the whole array into memory.

m-chrzan commented on packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
+    uint256 index,
+    uint256 blockNumber
+  ) public view onlyL2 returns (address) {
+    (, , , , address[] memory _elected) = getEpochManager().getEpochByBlockNumber(blockNumber);
```
Ditto

m-chrzan commented on packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
-    require(success, "error calling numberValidatorsInCurrentSet precompile");
-    return getUint256FromBytes(out, 0);
+    if (isL2()) {
+      return getEpochManager().getElected().length;
```
Ditto, implementing this in EpochManager should save gas since the whole `elected` array won't be read into memory.

soloseng replied:
The limitation in implementing this whole function in `EpochManager`, is to avoid importing the `UsingPrecompiles` contract.

I created an `EpochManager` function that returns the number instead of getting the elected array and getting its length.

m-chrzan commented on packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:
```diff
@@ -134,11 +179,17 @@ contract UsingPrecompiles is IsL2Check {
-    require(success, "error calling numberValidatorsInSet precompile");
-    return getUint256FromBytes(out, 0);
+    if (isL2()) {
+      (, , , , address[] memory elected) = getEpochManager().getEpochByBlockNumber(blockNumber);
```
Ditto

m-chrzan commented on packages/protocol/test-sol/utils08.sol:
```diff
@@ -20,12 +21,20 @@ contract Utils08 {
     vm.etch(0x4200000000000000000000000000000000000018, abi.encodePacked(bytes1(0x01)));
   }
+
+  // This function can be also found in OpenZeppelin's library, but in a newer version than the one
```
```suggestion
  // This function can be also found in OpenZeppelin's library, but in a newer version than the one we use.
```

m-chrzan reviewed (commented):
Looks good, but there's a few, potentially very impactful, gas optimizations possible. There's no need to read the entire `elected` array into memory when we only need one field of the Epoch struct, or just one of the elected addresses. These should be fairly simple changes that would reduce the number of `SLOAD`s by a factor of ~100 where relevant.

martinvol reviewed (changes requested):
I don't think we need to fix all precompiles in all contracts

martinvol commented on packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
-    require(success, "error calling validatorSignerAddressFromCurrentSet precompile");
-    return address(uint160(getUint256FromBytes(out, 0)));
+    if (isL2()) {
+      return getAccounts().getValidatorSigner(validatorAddressFromCurrentSet(index));
```
this wont work if signers change mid-epoch

martinvol commented on packages/protocol/contracts-0.8/common/UsingPrecompiles.sol:
```diff
@@ -73,59 +74,103 @@ contract UsingPrecompiles is IsL2Check {
+    if (isL2()) {
+      return getEpochManager().getEpochNumberOfBlock(blockNumber);
+    } else {
+      return epochNumberOfBlock(blockNumber, getEpochSize());
```
in L2 this is seconds instead of blocks

soloseng committed:
++ electedSigners

soloseng committed:
make usingPrecompile onlyL1 contract

soloseng committed:
++ precompile override contract

soloseng committed:
fix epoch manager bug and mock

soloseng committed:
using precompileOverride in validators contract

soloseng committed:
wording

soloseng committed:
using indexed function when querying specific validator accounts or signers on L2

soloseng commented:
Updated this PR based on our [last sync](https://github.com/celo-org/celo-blockchain-planning/issues/642#issuecomment-2397407994).

- added onlyL1 modifiers to all `UsingPrecompiles` functions. 
- created a new contract `PrecompileOverride` and it to handle the switching logic between L1 and L2. its currently being used by `Validators` contract but will need a 0.5.13 version for other core contracts still using 0.5.13 (e.g.: `Election`, Governance,EpochRewards

soloseng committed:
Merge branch 'release/core-contracts/12' into soloseng/UsingRegistry-backward-compatibility-fix

soloseng committed:
fixed failing test

soloseng committed:
using independent mapping for elected accounts and signers

soloseng committed:
removed old elected list

soloseng committed:
support 0.5.13 contracts using precompiles

soloseng commented:
As of commit 681a8582
Because we are wanting to keep a historical data of the elected validators account & signers, we now have to store the elected in a mapping. this causes an increase in gas when running the election, since we now have to write to 2 new storage slot (`electedAccountsOfEpoch` && `electedSignersOfEpoch` each time an election is run.

To minimize the gas cost of reading epoch info based on block number, the mapping of `electedAccountsOfEpoch` && `electedSignersOfEpoch` is kept separate than `epochs`. This allows us to retrieve the epoch info without reading the list of accounts and signers.

That being said, it would still be good to run a gas cost benchmark to know if it will be prohibitive.

usingPrecompile backwards compatibility has been added for `Election`, `Governance`, `EpochRewards` using a 0.5.13 version of the `PrecompileOverride` contract. This is because 0.5.13 does not support `virtual` && `override` function keywords.

soloseng committed:
-- registry import

pahor167 commented on packages/protocol/contracts/common/PrecompilesOverride.sol:
```diff
@@ -0,0 +1,112 @@
+   */
+  function getEpochNumberOfBlock(uint256 blockNumber) public view returns (uint256) {
+    if (isL2()) {
+      return getEpochManager().getEpochNumberOfBlock(blockNumber);
```
We currently dont have registry in this smart contract

soloseng committed:
++ registry in PrecompilesOverride

soloseng committed:
++ more tests

soloseng committed:
increase gaslimit

soloseng committed:
moved election history update to epoch processing start.

soloseng committed:
unused var

soloseng committed:
clean up

pahor167 commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -512,4 +658,86 @@ contract EpochManager is
+      electedSigners = new address[](_elected.length);
+    }
+
+    require(_elected.length == electedSigners.length, "Account and Signer list lenght mismatch.");
```
we cannot check whether _elected.length == electedSigners.length since in scenario when previously we had elected eg only 10 validators (because there were only 10 validators registered) and during this epoch new validators registered then it is valid to have different count of electedSigners from before

soloseng committed:
fixed e2e test

soloseng committed:
simplify code

soloseng committed:
Merge branch 'release/core-contracts/12' into soloseng/UsingRegistry-backward-compatibility-fix

martinvol commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -506,4 +649,82 @@ contract EpochManager is
+   * @return startTimestamp The starting timestamp of the given block number.
+   * @return rewardsBlock The reward block of the given block number.
+   */
+  function _getEpochByBlockNumber(
```
This function is very likely to run out of gas, is it (and the functions that use it) really needed?

I also think that it has quite some edge cases, would be difficult to test them all.

pahor167 replied:
comment added

martinvol committed:
Merge branch 'release/core-contracts/12' into soloseng/UsingRegistry-backward-compatibility-fix

soloseng committed:
Removed support for election historical data after that it's unlikely to be in use by external parties

pahor167 commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -51,13 +51,18 @@ contract EpochManager is
 
   EpochProcessState public epochProcessing;
   mapping(uint256 => Epoch) internal epochs;
+  mapping(uint256 => address[]) internal electedAccountsOfEpoch;
```
This variable doesnt seem to be set anywhere now

pahor167 commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -51,13 +51,18 @@ contract EpochManager is
+  mapping(uint256 => address[]) internal electedAccountsOfEpoch;
+  // Electeds in the L1 assumed signers can not change during the epoch
+  // so we keep a copy
+  mapping(uint256 => address[]) internal electedSignersOfEpoch;
```
Same here

pahor167 commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -51,13 +51,18 @@ contract EpochManager is
 
   uint256 public firstKnownEpoch;
   uint256 internal currentEpochNumber;
+  address[] internal electedAccounts;
```
Since this contract is already deployed on alfajores we should try not to break storage

soloseng replied:
made public

pahor167 commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -240,7 +251,7 @@ contract EpochManager is
         );
         processedGroups[group] = epochRewards == 0 ? type(uint256).max : epochRewards;
       }
-      delete elected[i];
```
We should be deleting also signers here

pahor167 commented on packages/protocol/contracts-0.8/common/EpochManager.sol:
```diff
@@ -264,8 +275,17 @@ contract EpochManager is
+    _setElectedSigners(_newlyElected);
+
+    EpochProcessState memory _epochProcessingEmpty;
+    epochProcessing = _epochProcessingEmpty;
```
Do we need to zero this out here ?

soloseng replied:
This was a suggestion from @martinvol. I don't have any issues with zeroing it out at the end of the epoch process.

pahor167 replied:
leaving as it is now

soloseng committed:
PR feedback

soloseng committed:
-- spacing

soloseng committed:
reorder state var

soloseng committed:
stat var reorder

soloseng committed:
small fix

soloseng committed:
delete electedSigners at end of epochprocessing

soloseng committed:
CI fix

pahor167 committed:
added comment
